numba/core/transforms.py

```diff
@@ -477,73 +476,126 @@ def _cfg_nodes_in_region(cfg, region_begin, region_end):
+                for stmt in blocks[block].body:
+                    # raise detected before pop_block
+                    if is_raise(stmt):
+                            raise errors.CompilerError(
+                                'unsupported controlflow due to raise '
+                                'statements inside with block'
+                                )
+                    # special case 3.7, return before POP_BLOCK
+                    if PYVERSION < (3, 8) and is_return(stmt):
+                            raise errors.CompilerError(
+                                'unsupported controlflow due to return '
+                                'statements inside with block'
+                                )
+                    # if a pop_block, process it
+                    if is_pop_block(stmt) and block in sus_pops:
+                        # record the jump target of this block belonging to this setup
+                        pop_block_targets = blocks[block].terminator.get_targets()
+                        if len(pop_block_targets) != 1:
+                            raise errors.CompilerError(
+                                "Does not support with-context that contain branches "
+                                "(i.e. break/return/raise) that can leave the with-context. "
+                            )
+                        target_block = blocks[pop_block_targets[0]]
+                        if not target_block.terminator.get_targets():
+                            raise errors.CompilerError(
+                                'unsupported controlflow due to return '
+                                'statements inside with block'
+                                )
+                        setup_with_to_pop_blocks_map[setup_block].add(pop_block_targets[0])
+                        # remove the block from blocks to be matched
+                        sus_pops.remove(block)
+                        # stop looking, we have reached the frontier
+                        break
+                    # if we are still here, by the terminator block,
+                    # add all it's targets to the to_visit stack, unless we
+                    # have seen them already
+                    if is_terminator(stmt):
+                        for t in stmt.get_targets():
+                            if t not in seen:
+                                to_visit.append(t)
+
+        for setup_with, pop_blocks in setup_with_to_pop_blocks_map.items():
+            for p in pop_blocks:
+                yield setup_with, p
+
+
+    known_ranges = []
+
+    def within_known_range(start, end, known_ranges):
         for a, b in known_ranges:
-            if start >= a and start < b:
+            # FIXME: this should be a comparison in topological order, right
+            # now we are comparing the integers of the blocks, stuff probably
+            # works by accident.
+            if start > a and end < b:
```

Comment:
This is now strictly `>` and `<`, i.e. no equality, is this because of a change in the block emission, i.e. starts and ends have to be in their own unique blocks?

Reply:
Yes, it is a bit contrived. The previous representation checked if `start` was between `a` and `b`. Because having a `SETUP_WITH` inside the bounds (as determined by the offset argument to `SETUP_WITH`). Now that we have the exact locations of `SETUP_WITH` AND `POP_BLOCK` we can check that a given `SETUP_WITH`/`POP_BLOCK` pair  is contained entirely within the bounds of an enclosing pair, i.e. nested. 

Anyway, to answer the original question, we never have duplicate `SETUP_WITH`s or `POP_BLOCKS` because there will only ever be one per block. This means, we can use strict comparison.

Reply:
Thanks for the explanation, this makes sense with respect to the proposed changes.